src/chainstate/burn/operations/stack_stx.rs

```diff
@@ -0,0 +1,495 @@
+        tx: &BurnchainTransaction,
+        pox_sunset_ht: u64,
+    ) -> Result<PreStackStxOp, op_error> {
+        // can't be too careful...
+        let inputs = tx.get_signers();
+        let outputs = tx.get_recipients();
+
+        if inputs.len() == 0 {
+            warn!(
+                "Invalid tx: inputs: {}, outputs: {}",
+                inputs.len(),
+                outputs.len()
+            );
+            return Err(op_error::InvalidInput);
+        }
+
+        if outputs.len() == 0 {
+            warn!(
+                "Invalid tx: inputs: {}, outputs: {}",
+                inputs.len(),
+                outputs.len()
+            );
+            return Err(op_error::InvalidInput);
+        }
+
+        if tx.opcode() != Opcodes::PreStackStx as u8 {
+            warn!("Invalid tx: invalid opcode {}", tx.opcode());
+            return Err(op_error::InvalidInput);
+        };
+
+        // check if we've reached PoX disable
+        if block_height >= pox_sunset_ht {
+            debug!(
+                "PreStackStxOp broadcasted after sunset. Ignoring. txid={}",
+                tx.txid()
+            );
+            return Err(op_error::InvalidInput);
+        }
+
+        Ok(PreStackStxOp {
+            output: outputs[0].address,
+            txid: tx.txid(),
+            vtxindex: tx.vtxindex(),
+            block_height,
+            burn_header_hash: block_hash.clone(),
+        })
+    }
+}
+
+impl StackStxOp {
+    #[cfg(test)]
+    pub fn new(
+        sender: &StacksAddress,
+        reward_addr: &StacksAddress,
+        stacked_ustx: u128,
+        num_cycles: u8,
+    ) -> StackStxOp {
+        StackStxOp {
+            sender: sender.clone(),
+            reward_addr: reward_addr.clone(),
+            stacked_ustx,
+            num_cycles,
+            // to be filled in
+            txid: Txid([0u8; 32]),
+            vtxindex: 0,
+            block_height: 0,
+            burn_header_hash: BurnchainHeaderHash([0u8; 32]),
+        }
+    }
+
+    fn parse_data(data: &Vec<u8>) -> Option<ParsedData> {
+        /*
+            Wire format:
+            0      2  3                             19        20
+            |------|--|-----------------------------|---------|
+             magic  op         uSTX to lock (u128)     cycles (u8)
+
+             Note that `data` is missing the first 3 bytes -- the magic and op have been stripped
+
+             The values ustx to lock and cycles are in big-endian order.
+
+             parent-delta and parent-txoff will both be 0 if this block builds off of the genesis block.
+        */
+
+        if data.len() < 17 {
+            // too short
+            warn!(
+                "StacksStxOp payload is malformed ({} bytes, expected {})",
+                data.len(),
+                17
+            );
+            return None;
+        }
+
+        let stacked_ustx = parse_u128_from_be(&data[0..16]).unwrap();
+        let num_cycles = data[16].clone();
```

Comment:
Is the `.clone()` necessary?  Doesn't `u8` implement `Copy`?

Reply:
Ah yeah, this clone() can be removed.

Reply:
Updated in 74da104fb